File: dojo/src/lib/server/forge-refresh.ts
// Redeem a GitHub refresh token for a live access token.
//
// ## Why this lives in the dōjō and not in the daemon
//
// Redeeming a refresh token requires the OAuth app's CLIENT SECRET. The daemon
// runs on a user's laptop; shipping the secret there would put it on every
// machine that installs sensei, where a single extracted copy would let anyone
// mint tokens against the app. So the daemon holds the refresh token — which is
// scoped to one user and revocable — and the dōjō holds the secret. Neither can
// refresh alone, which is the point.
//
// ## Measured, not assumed
//
// Verified against the real app before this module existed: a redemption returns
// `expires_in: 28800` (8h) and `refresh_token_expires_in: 15724800` (182d), with
// `scope` preserved. The 8h lifetime is why a token signed in yesterday is dead
// this morning, and why `sensei.personas.forge_token_expires_at` is worth
// recording at all.
//
// PURE: every dependency is injected, so this is testable without the
// `$env/dynamic/private` virtual module (which vitest does not generate). The
// env-reading glue is the route.
import { AdminError } from './admin-data';

/** GitHub's OAuth token endpoint. Not the API host — `api.github.com` answers
 *  404 for this path, which reads as "no such user" rather than "wrong host". */
const TOKEN_URL = 'https://github.com/login/oauth/access_token';

/** The OAuth app's identity. Never logged, never returned, never in an error. */
export interface ForgeAppCredentials {
	clientId: string;
	clientSecret: string;
}

/** A live forge credential, with an ABSOLUTE deadline. */
export interface ForgeRefreshResult {
	accessToken: string;
	/** The refresh token to store now. GitHub rotates on every redemption, so
	 *  this is usually NEW — but it falls back to the one that was sent, because
	 *  a response without one means the old token still stands. */
	refreshToken: string;
	/** Unix seconds. Absolute, because the daemon reads it much later than the
	 *  response that produced it. */
	expiresAt: number;
	/** What the token can do, as GitHub reports it. Worth returning: a refresh
	 *  that silently narrows scope explains an authorization failure that would
	 *  otherwise look like a bug. */
	scope: string | null;
}

/**
 * A refresh that did not produce a token.
 *
 * `needsSignIn` is the load-bearing field, and it is the same distinction the
 * daemon draws for the session token: a REJECTED grant is terminal and only the
 * user can fix it; an outage is not, and telling someone to re-authenticate for
 * a 502 makes them destroy a working session for nothing.
 */
export class ForgeRefreshError extends AdminError {
	constructor(
		message: string,
		readonly needsSignIn: boolean
	) {
		// 502: this endpoint is a proxy for GitHub, and the failure is GitHub's.
		super(502, message);
	}
}

/** GitHub's terminal refusals. Anything else — a 5xx, a timeout, an unparseable
 *  body — is treated as transient, because the cost of guessing wrong in that
 *  direction is a needless sign-out. */
const TERMINAL = new Set([
	'bad_refresh_token',
	'bad_verification_code',
	'incorrect_client_credentials',
	'unauthorized_client',
	'invalid_grant',
	'access_denied'
]);

export interface RefreshDeps {
	fetchImpl?: typeof fetch;
	/** Unix seconds. Injected so the absolute expiry is assertable. */
	now?: () => number;
}

/**
 * Exchange a refresh token for a live access token.
 *
 * Throws [`ForgeRefreshError`] rather than returning a partial result: a caller
 * that cannot tell success from failure will write whatever it got into the
 * Keychain, and an empty string stored over a working credential is a sign-out
 * with no explanation.
 */
export async function refreshForgeToken(
	refreshToken: string,
	creds: ForgeAppCredentials,
	deps: RefreshDeps = {}
): Promise<ForgeRefreshResult> {
	const fetchImpl = deps.fetchImpl ?? fetch;
	const now = deps.now ?? (() => Math.floor(Date.now() / 1000));

	const res = await fetchImpl(TOKEN_URL, {
		method: 'POST',
		// Without this GitHub answers form-encoded and `res.json()` throws on a
		// response that actually succeeded.
		headers: { accept: 'application/json', 'content-type': 'application/x-www-form-urlencoded' },
		body: new URLSearchParams({
			grant_type: 'refresh_token',
			refresh_token: refreshToken,
			client_id: creds.clientId,
			client_secret: creds.clientSecret
		}).toString()
	});

	// Parsed before `res.ok` is consulted, because the failure mode that matters
	// arrives as HTTP 200.
	const body = (await res.json().catch(() => null)) as Record<string, unknown> | null;

	if (!res.ok) {
		throw new ForgeRefreshError(`GitHub refused the refresh (HTTP ${res.status})`, false);
	}
	if (!body) {
		throw new ForgeRefreshError('GitHub returned an unreadable refresh response', false);
	}

	// The OAuth quirk this module exists to survive: a refusal is an HTTP 200
	// with an `error` field. Trusting `res.ok` alone yields `accessToken:
	// undefined` reported as a success.
	const error = typeof body.error === 'string' ? body.error : null;
	if (error) {
		// The code only — never `error_description`, which echoes the request and
		// has been observed to quote the credential that was sent.
		throw new ForgeRefreshError(`GitHub refused the refresh: ${error}`, TERMINAL.has(error));
	}

	const accessToken = typeof body.access_token === 'string' ? body.access_token : '';
	if (!accessToken) {
		// A 200 with neither an error nor a token is a shape we do not
		// understand. Not terminal: it is far more likely a proxy or an outage
		// than a revoked grant, and a sign-out is the expensive guess.
		throw new ForgeRefreshError('GitHub returned no access token and no error', false);
	}

	const expiresIn = typeof body.expires_in === 'number' ? body.expires_in : null;
	return {
		accessToken,
		// Falls back to the token we sent. GitHub rotates, but a response without
		// a new one means the old is still valid, and erasing it would destroy
		// the only credential that can recover the session — after a SUCCESS.
		refreshToken: typeof body.refresh_token === 'string' ? body.refresh_token : refreshToken,
		// No `expires_in` means no deadline was stated. Inventing one would make
		// the daemon schedule a refresh against a fabricated time; `0` is
		// distinguishable and the caller records it as unknown.
		expiresAt: expiresIn === null ? 0 : now() + expiresIn,
		scope: typeof body.scope === 'string' ? body.scope : null
	};
}

File: crates/senseid/src/dojo_client/forge_refresh.rs
//! Renew a forge token through the dōjō.
//!
//! The daemon holds the refresh token and cannot spend it: redemption needs the
//! GitHub App's client secret, which lives in the dōjō and deliberately never
//! ships to a user's machine. So this sends the refresh token to
//! `POST /v1/you/forge/refresh` and stores what comes back.
//!
//! ## Why the outcome has three arms and not two
//!
//! The same split [`crate::api::handlers::auth::AuthError`] draws for the
//! session token, for the same reason. A REJECTED refresh is terminal — the
//! grant is gone and only the user can restore it — while an outage is not, and
//! treating the two alike either nags about a blip or stays silent about a grant
//! that will never work again. The dōjō returns `needsSignIn` as a field so this
//! never has to parse English to tell them apart.
//!
//! ## The rotation is the dangerous part
//!
//! GitHub issues a NEW refresh token on every successful redemption and
//! invalidates the old one. If the new token is not stored, the session is
//! unrecoverable without a sign-in — so the write happens before the caller is
//! told anything succeeded, and a failed write is reported as a failure.

use crate::dojo_client::session;
use crate::dojo_client::settings::dojo_url;

/// What a renewal attempt produced.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RefreshOutcome {
    /// A live token, already stored. `expires_at` is unix seconds, `0` when the
    /// dōjō reported no deadline — distinguishable from a real one, and never
    /// invented.
    Renewed { expires_at: i64 },
    /// The grant is gone. Only a sign-in restores it; retrying cannot.
    Rejected(String),
    /// We could not complete the exchange. Says NOTHING about the token — the
    /// stored credential must be left exactly as it was.
    Unavailable(String),
}

/// The new credentials, before they are stored.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Renewed {
    pub access_token: String,
    pub refresh_token: String,
    pub expires_at: i64,
}

/// Classify the dōjō's answer.
///
/// Split from the I/O so every branch is testable without a dōjō. `status` is
/// `None` when no response arrived at all.
pub fn classify_refresh(status: Option<u16>, body: &str) -> Result<Renewed, RefreshOutcome> {
    let Some(status) = status else {
        return Err(RefreshOutcome::Unavailable("no response from the dōjō".into()));
    };
    let parsed: Option<serde_json::Value> = serde_json::from_str(body).ok();

    if !(200..300).contains(&status) {
        // `needsSignIn` is a FIELD, not a status code: a 502 carries both the
        // terminal and the transient case, because both are GitHub refusing.
        let terminal = parsed.as_ref().and_then(|v| v["needsSignIn"].as_bool()).unwrap_or(false);
        let msg = parsed
            .as_ref()
            .and_then(|v| v["error"].as_str())
            .map(str::to_owned)
            .unwrap_or_else(|| format!("dōjō refused the refresh (HTTP {status})"));
        // Absent or unreadable `needsSignIn` is treated as TRANSIENT. Guessing
        // terminal signs the user out over a proxy error; guessing transient
        // costs one wasted retry on the next interval.
        return Err(if terminal {
            RefreshOutcome::Rejected(msg)
        } else {
            RefreshOutcome::Unavailable(msg)
        });
    }

    let Some(v) = parsed else {
        return Err(RefreshOutcome::Unavailable("dōjō returned an unreadable response".into()));
    };
    let access_token = v["access_token"].as_str().unwrap_or_default().to_string();
    let refresh_token = v["refresh_token"].as_str().unwrap_or_default().to_string();
    // A 2xx with no token is a shape we do not understand. Storing the empty
    // string would overwrite a working credential with nothing, and every later
    // call would 401 with no record of why.
    if access_token.is_empty() {
        return Err(RefreshOutcome::Unavailable("dōjō returned no access token".into()));
    }
    if refresh_token.is_empty() {
        return Err(RefreshOutcome::Unavailable("dōjō returned no refresh token".into()));
    }
    Ok(Renewed { access_token, refresh_token, expires_at: v["expires_at"].as_i64().unwrap_or(0) })
}

/// Renew one persona's forge token, storing the result.
///
/// Both writes are attempted and BOTH are load-bearing. The refresh token is the
/// one that cannot be recovered: GitHub already invalidated the old one when it
/// issued this, so failing to store it silently would leave a session that dies
/// at the next expiry with no way back except a sign-in.
pub async fn refresh_persona(persona: &str) -> RefreshOutcome {
    let who = persona.to_string();
    let stored =
        tokio::task::spawn_blocking(move || session::load_provider_refresh_token(&who)).await;
    let refresh_token = match stored {
        Ok(Ok(t)) => t,
        // No refresh token means renewal was never possible for this persona —
        // a session predating provider-refresh capture, or a forge that issued
        // none. Not a fault to retry.
        _ => {
            return RefreshOutcome::Rejected(
                "no forge refresh token is stored for this persona".into(),
            );
        }
    };

    // The dōjō authenticates the caller before it spends anything, so this needs
    // a live SESSION token — a different credential from the one being renewed.
    let bearer = match crate::api::handlers::auth::live_access_token(persona).await {
        Ok(t) => t,
        Err(e) => {
            return RefreshOutcome::Unavailable(format!("no dōjō session to refresh with: {e:?}"));
        }
    };

    let resp = crate::federation::http_client()
        .post(format!("{}/v1/you/forge/refresh", dojo_url()))
        .bearer_auth(bearer)
        .json(&serde_json::json!({ "refresh_token": refresh_token }))
        .send()
        .await;

    let (status, body) = match resp {
        Ok(r) => {
            let s = r.status().as_u16();
            (Some(s), r.text().await.unwrap_or_default())
        }
        Err(e) => {
            return RefreshOutcome::Unavailable(format!("could not reach the dōjō: {e}"));
        }
    };

    let renewed = match classify_refresh(status, &body) {
        Ok(r) => r,
        Err(outcome) => return outcome,
    };

    let (at, rt) = (renewed.access_token.clone(), renewed.refresh_token.clone());
    let who = persona.to_string();
    let written = tokio::task::spawn_blocking(move || {
        session::store_provider_token(&who, &at)?;
        session::store_provider_refresh_token(&who, &rt)
    })
    .await;
    // Reported as a failure, loudly. GitHub has already invalidated the old
    // refresh token, so a persona whose new one did not store now holds a
    // credential nothing can renew — the user must sign in, and answering
    // "renewed" would hide that until the token dies hours later.
    let detail = match written {
        Ok(Ok(())) => return RefreshOutcome::Renewed { expires_at: renewed.expires_at },
        Ok(Err(e)) => e.to_string(),
        Err(e) => format!("the keychain write did not complete: {e}"),
    };
    tracing::error!(persona, error = %detail,
                    "forge token renewed but could NOT be stored — the old token is already \
                     invalid, so this persona needs a sign-in");
    RefreshOutcome::Unavailable(format!("renewed but could not store: {detail}"))
}

#[cfg(test)]
mod refresh_classification {
    use super::*;

    const OK: &str = r#"{"access_token":"gho_new","refresh_token":"ghr_new",
                         "expires_at":1788148800,"scope":"read:org,repo"}"#;

    #[test]
    fn a_renewal_carries_both_tokens_and_the_absolute_deadline() {
        let r = classify_refresh(Some(200), OK).expect("a 200 with both tokens is a renewal");
        assert_eq!(r.access_token, "gho_new");
        assert_eq!(r.refresh_token, "ghr_new");
        assert_eq!(r.expires_at, 1_788_148_800);
    }

    #[test]
    fn a_refusal_that_says_it_needs_a_sign_in_is_terminal() {
        // The poison-pill guard. A revoked grant fails identically forever;
        // without this the scheduled check retries it every interval until the
        // user happens to sign in on their own.
        let body =
            r#"{"error":"GitHub refused the refresh: bad_refresh_token","needsSignIn":true}"#;
        assert_eq!(
            classify_refresh(Some(502), body),
            Err(RefreshOutcome::Rejected("GitHub refused the refresh: bad_refresh_token".into()))
        );
    }

    #[test]
    fn a_refusal_without_that_flag_is_treated_as_transient() {
        // Deliberately asymmetric. Guessing TERMINAL on a proxy error signs the
        // user out of a session that was fine; guessing transient costs one
        // wasted retry on the next interval. Only one of those is recoverable.
        let body = r#"{"error":"GitHub refused the refresh (HTTP 502)","needsSignIn":false}"#;
        assert!(matches!(classify_refresh(Some(502), body), Err(RefreshOutcome::Unavailable(_))));
        // And when the field is missing entirely.
        assert!(matches!(
            classify_refresh(Some(500), r#"{"error":"boom"}"#),
            Err(RefreshOutcome::Unavailable(_))
        ));
        // And when the body is not JSON at all — an HTML error page from a proxy.
        assert!(matches!(
            classify_refresh(Some(504), "<html>gateway timeout</html>"),
            Err(RefreshOutcome::Unavailable(_))
        ));
    }

    #[test]
    fn an_unconfigured_dojo_is_transient_not_a_revoked_grant() {
        // 503 from `forgeAppFromEnv() == null`. The deployment is missing its
        // client secret; telling the user to sign in again cannot fix that, and
        // they would destroy a working session trying.
        let body = r#"{"error":"this dōjō is not configured to refresh forge tokens"}"#;
        assert!(matches!(classify_refresh(Some(503), body), Err(RefreshOutcome::Unavailable(_))));
    }

    #[test]
    fn no_response_at_all_says_nothing_about_the_token() {
        assert!(matches!(classify_refresh(None, ""), Err(RefreshOutcome::Unavailable(_))));
    }

    #[test]
    fn a_success_missing_either_token_is_refused_rather_than_stored() {
        // Storing an empty string would overwrite a working credential with
        // nothing; every later call then 401s with no record of why. And a
        // missing REFRESH token is worse than it looks — GitHub has already
        // invalidated the old one by the time it answers, so accepting this
        // leaves a session that cannot be renewed again.
        for body in [
            r#"{"refresh_token":"ghr_new","expires_at":1}"#,
            r#"{"access_token":"gho_new","expires_at":1}"#,
            r#"{"access_token":"","refresh_token":"","expires_at":1}"#,
        ] {
            assert!(
                matches!(classify_refresh(Some(200), body), Err(RefreshOutcome::Unavailable(_))),
                "accepted a partial renewal: {body}"
            );
        }
    }

    #[test]
    fn a_renewal_with_no_stated_deadline_records_zero_rather_than_inventing_one() {
        // `0` is distinguishable and the caller stores it as unknown. Making up
        // "now + 8h" would schedule the next renewal against a fabricated time.
        let r = classify_refresh(Some(200), r#"{"access_token":"a","refresh_token":"b"}"#)
            .expect("a renewal without a deadline is still a renewal");
        assert_eq!(r.expires_at, 0);
    }
}
